src/main/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/jpa/specification/UserSpecification.kt

```diff
@@ -0,0 +1,40 @@
+import org.springframework.data.jpa.domain.Specification
+import uk.gov.justice.digital.hmpps.nomisuserrolesapi.jpa.UserGroup
+import uk.gov.justice.digital.hmpps.nomisuserrolesapi.jpa.UserGroupAdministrator
+import uk.gov.justice.digital.hmpps.nomisuserrolesapi.jpa.UserGroupAdministratorPk
+import uk.gov.justice.digital.hmpps.nomisuserrolesapi.jpa.UserGroupMember
+import uk.gov.justice.digital.hmpps.nomisuserrolesapi.jpa.UserPersonDetail
+import javax.persistence.criteria.CriteriaBuilder
+import javax.persistence.criteria.CriteriaQuery
+import javax.persistence.criteria.Predicate
+import javax.persistence.criteria.Root
+
+class UserSpecification(private val filter: UserFilter) : Specification<UserPersonDetail> {
```

Comment:
There are two things that would make this work so much better

1. org.hibernate:hibernate-jpamodelgen that generates a type safe model - but Kotlin annotation processing is broken in Java 16/17 
2. https://github.com/consoleau/kotlin-jpa-specification-dsl - a Kotlin DSL - but there are shenanigans about this not being in the Mavern repo and the author has left the company that wrote it 

So at least ended using Kotlin so the property names are not hardcode so refactoring would work.

Ugly code though